---
title: Error Handling
description: ShipEngine Connect applications can throw errors just like any other JavaScript application. We also support custom error properties and classes to make things easier.
---



Error Handling
=================

ShipEngine Connect applications can throw [Errors](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) just like any other JavaScript application.  Any uncaught exception will terminate the currently executing method and generate an error log.

For example, the methods you implement will often need to perform input validation. If this validation fails, you can throw an error as in the example below.

```javascript highlight="3"
for (let parcel of shipment.packages) {
  if (parcel.packaging.code === "BOX" && parcel.weight.grams > 100000) {
    throw new RangeError(`${parcel.packaging.name} cannot weigh more than 100 kilograms`);
  }
}
```

> **TIP:** In addition to [JavaScript's built-in error classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Error_objects), ShipEngine Connect includes several [error classes](#error-classes) you can use in your app.


<Section>
<Left>
Additional Error Properties
----------------------------------------
Before throwing an error, you can set additional properties that provide more context, enabling our platform to respond appropriately and provide meaningful feedback to users.

<Reference>
  <Field name="code" type="string">
    <Description>
      A string that identifies the type of error. Most errors that occur in Node.js already have a code, but if you're throwing a new error, then we recommend that you use one of our pre-defined [Error Codes](#error-codes) below
    </Description>
  </Field>

  <Field name="statusCode" type="number">
    <Description>
      The numeric status code associated with the error, if any. For errors that originate from an HTTP response, this should be the HTTP status code (e.g. 400, 404, 500)
    </Description>
  </Field>

  <Field name="externalErrors" type="string[]">
    <Description>
      If the error is the result of calling an external service, then put any error messages that were returned by the service here.
    </Description>
  </Field>

  <Field name="externalWarnings" type="string[]">
    <Description>
      If the error is the result of calling an external service, then put any warning messages that were returned by the service here.
    </Description>
  </Field>

  <Field name="originalError">
    <Type>
      [Error](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error)
    </Type>
    <Description>
      The original error that occurred, if any. This is especially useful for debugging, since it will allow you to see the error message and stack trace where the problem occurred.
    </Description>
  </Field>

  <Field name="*" type="any value">
    <Description>
      Any additional properties you add to the error will be logged, which is useful for debugging.
    </Description>
  </Field>
</Reference>


Error Codes
---------------------
Most errors that occur in Node.js have a `code` property. When throwing your own errors, we encourage you to include a `code` as well. Using one of the following codes will help the ShipEngine Connect platform know how to handle your error appropriately.

<Reference>
  <Field name="ERR_INVALID">
    <Description>
      You should use this code to indicate a validation error. For example, your app may only allow certain values under specific conditions, or a field may be conditionally required.
    </Description>
  </Field>

  <Field name="ERR_UNAUTHORIZED">
    <Description>
      Use this code to indicate an authentication or authorization error. For example, if the `transaction.session` object is missing some required data or the session has expired.
    </Description>
  </Field>

  <Field name="ERR_EXTERNAL">
    <Description>
      You should use this code for any error that is the result of calling an external service, such as an API. This lets ShipEngine Connect know that the error came from a downstream service rather than directly in your app.

      You should also set the `statusCode` property if possible to indicate the type of external error that occurred. For HTTP errors, you can use the HTTP status code (e.g. 400, 404, 500). For APIs that don't use HTTP status codes, we recomend using the most appropriate [HTTP status code](https://httpstatuses.com/) and also put the raw error/warning messages in `externalErrors` or `externalWarnings`.
    </Description>
  </Field>

  <Field name="ERR_APP_ERROR">
    <Description>
      This is a generic error code that simply indicates that the error was _intentionally_ thrown by your app, as opposed to an unexpected error that occurred due to a bug or runtime exception.
    </Description>
  </Field>
</Reference>
</Left>
<Right>
### Validation Error Example
This example demonstrates throwing a validation error with additional properties.

```javascript
if (package.weight.grams > 100000) {
  throw Object.assign(
    new Error("Weight cannot exceed 100 kilograms"),
    {
      code: "ERR_INVALID",
      actualWeight: package.weight.grams,
      maxWeight: 100000,
    }
  );
}
```

### External Error Example
This example rethrows an error from an external service, but first it adds additional properties.

```javascript
try {
  // Some code that makes an HTTP call
}
catch (error) {
  error.code = "ERR_EXTERNAL";
  error.statusCode = httpResponse.status;
  error.externalErrors = [httpResponse.body.errorMessage];
  error.headers = httpResponse.headers;
  throw error;
}
```
</Right>
</Section>


Error Classes
--------------------------
In addition to [JavaScript's built-in error classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects#Error_objects), ShipEngine Connect includes several [error classes](#error-classes) you can use in your app. Each class corresponds to one of the [error codes](#error-codes) listed above and automatically sets the `code` property for you. You can also pass [additional error properties](#additional-error-properties) to the constructor.

<Section>
<Left>
### `ValidationError`
You should use this class for validation errors. For example, your app may only allow certain values under specific conditions, or a field may be conditionally required.

#### `constructor(message | props)`
The `ValidationError` constructor can accept a string (the error message) or an object containing [additional error properties](#additional-error-properties).
</Left>
<Right>
<CodeWrapper>
```javascript
const { ValidationError } = require("@shipengine/connect");

// This only sets the `code` and `message` properties
throw new ValidationError("The package weight exceeds the maximum allowed");

// This sets the `code`, `message`, and additional properties
throw new ValidationError({
  message: "The package weight exceeds the maximum allowed",
  actualWeight: package.weight.ounces,
  maxWeight: 2400,
});
```

```typescript
import { ValidationError } from "@shipengine/connect";

// This only sets the `code` and `message` properties
throw new ValidationError("The package weight exceeds the maximum allowed");

// This sets the `code`, `message`, and additional properties
throw new ValidationError({
  message: "The package weight exceeds the maximum allowed",
  actualWeight: package.weight.ounces,
  maxWeight: 2400,
});
```
</CodeWrapper>
</Right>
</Section>

<Section>
<Left>
### `UnauthorizedError`
You should use this class to indicate an authentication or authorization error. For example, if the `transaction.session` object is missing some required data or the session has expired.

#### `constructor(message | props)`
The `UnauthorizedError` constructor can accept a string (the error message) or an object containing [additional error properties](#additional-error-properties).
</Left>
<Right>
<CodeWrapper>
```javascript
const { UnauthorizedError } = require("@shipengine/connect");

// This only sets the `code` and `message` properties
throw new UnauthorizedError("Your session has expired");

// This sets the `code`, `message`, and additional properties
throw new UnauthorizedError({
  message: "Your session has expired",
  sessionID: session.id,
  expiredAt: ession.expireDateTime,
});
```

```typescript
import { UnauthorizedError } from "@shipengine/connect";

// This only sets the `code` and `message` properties
throw new UnauthorizedError("Your session has expired");

// This sets the `code`, `message`, and additional properties
throw new UnauthorizedError({
  message: "Your session has expired",
  sessionID: session.id,
  expiredAt: ession.expireDateTime,
});
```
</CodeWrapper>
</Right>
</Section>

<Section>
<Left>
### `ExternalError`
You should use this class when the error is the result of calling an external service, such as an API. This lets ShipEngine Connect know that the error came from a downstream service rather than directly in your app.

You should also set the `statusCode` property if possible to indicate the type of external error that occurred. For HTTP errors, you can use the HTTP status code (e.g. 400, 404, 500). For APIs that don't use HTTP status codes, we recomend using the most appropriate [HTTP status code](https://httpstatuses.com/) and also put the raw error/warning messages in `externalErrors` or `externalWarnings`.

#### `constructor(message | props)`
The `ExternalError` constructor can accept a string (the error message) or an object containing [additional error properties](#additional-error-properties).
</Left>
<Right>
<CodeWrapper>
```javascript
const { ExternalError } = require("@shipengine/connect");

// This only sets the `code` and `message` properties
throw new ExternalError("The request exceeded the rate limit");

// This sets the `code`, `message`, and additional properties
throw new ExternalError({
  message: "The request exceeded the rate limit",
  statusCode: 429,
  externalErrors: [httpResponse.body.errorMessage],
});
```

```typescript
import { ExternalError } from "@shipengine/connect";

// This only sets the `code` and `message` properties
throw new ExternalError("The request exceeded the rate limit");

// This sets the `code`, `message`, and additional properties
throw new ExternalError({
  message: "The request exceeded the rate limit",
  statusCode: 429,
  externalErrors: [httpResponse.body.errorMessage],
});
```
</CodeWrapper>
</Right>
</Section>

<Section>
<Left>
### `AppError`
This is a generic error class that simply distinguishes errors that are _intentionally_ thrown by your app, as opposed to an unexpected error that occurred due to a bug or runtime exception.

#### `constructor(message | props)`
The `AppError` constructor can accept a string (the error message) or an object containing [additional error properties](#additional-error-properties).
</Left>
<Right>
<CodeWrapper>
```javascript
const { AppError } = require("@shipengine/connect");

// This only sets the `code` and `message` properties
throw new AppError("Unable to load app data");

// This sets the `code`, `message`, and additional properties
throw new AppError({
  message: "Unable to load app data",
  dataSource: "./app-data.xml",
});
```

```typescript
import { AppError } from "@shipengine/connect";

// This only sets the `code` and `message` properties
throw new AppError("Unable to load app data");

// This sets the `code`, `message`, and additional properties
throw new AppError({
  message: "Unable to load app data",
  dataSource: "./app-data.xml",
});
```
</CodeWrapper>
</Right>
</Section>



<Pager
  prev="./typescript.mdx" prevTitle="Developing in TypeScript"
  next="./testing/index.mdx" nextTitle="Testing"
/>
